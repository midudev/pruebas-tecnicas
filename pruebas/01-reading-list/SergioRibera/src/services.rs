mod data;

pub use data::*;
use serde::de::DeserializeOwned;

use crate::errors::Error;
#[cfg(feature = "inspect")]
use log::error;

pub async fn get<T>(url: String) -> Result<T, Error>
where
    T: DeserializeOwned,
{
    let response = reqwest::get(url).await;
    match response {
        Ok(data) => (data.json::<T>().await).map_or(Err(Error::DeserializeError), |repo| Ok(repo)),
        Err(_e) => {
            #[cfg(feature = "inspect")]
            error!("Loaded Data Failed: {_e:?}");
            Err(Error::RequestError)
        }
    }
}
