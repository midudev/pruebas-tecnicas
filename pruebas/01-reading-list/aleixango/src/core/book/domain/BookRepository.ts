import { Book } from "./Book"

export interface BookRepository {
  get(filterGenre: string): Promise<Book[]>
  add(book: Book): Promise<boolean>
  remove(isbn: string): Promise<boolean>
}
