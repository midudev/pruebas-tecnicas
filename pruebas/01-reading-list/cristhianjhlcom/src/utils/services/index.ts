export { default as LocalStorageService } from './LocalStorageService';
