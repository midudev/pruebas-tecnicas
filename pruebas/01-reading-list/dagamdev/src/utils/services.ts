import books from './books.json'
import { LOCAL_KEYS } from './config'
import { Book, LocalKeys } from './types'

export function getBooks() {
  return books.library.map(({book})=> book)
}

export function getBook(bookISBN: string) {
  return books.library.find(f=> f.book.ISBN == bookISBN)?.book
}

export const windowExist = typeof window != 'undefined'

export function getBooksISBNs(localKey: LocalKeys) {
  if(typeof localStorage != 'undefined'){
    const localISBNsData = localStorage.getItem(localKey)
    return localISBNsData ? JSON.parse(localISBNsData) as string[] : []
  }

  return []
}

export function getBooksByISBNs(ISBNs: string[]) {
  return getBooks().filter(f=> ISBNs.some((s: string)=> s == f.ISBN))
}

function getBooksFromLocalStorage(localKey: LocalKeys) {
  const readingBooksISBNs = getBooksISBNs(localKey)
  
  if(Array.isArray(readingBooksISBNs)) return getBooksByISBNs(readingBooksISBNs)
  
  return []
}

export function getReadingBooks() {
  return getBooksFromLocalStorage(LOCAL_KEYS.READING_BOOKS)
}

export function getPriorityBooks() {
  return getBooksFromLocalStorage(LOCAL_KEYS.PRIORITY_BOOKS)
}

export function updateReadingBooks(booksISBNs: Book[]) {
  localStorage.setItem(LOCAL_KEYS.READING_BOOKS, JSON.stringify(booksISBNs.map(m=> m.ISBN)))
}