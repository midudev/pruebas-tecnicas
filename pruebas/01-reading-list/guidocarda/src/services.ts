import data from "./books.json";

export const getBooks = () => {
  return data.library.map((item) => item.book);
};
