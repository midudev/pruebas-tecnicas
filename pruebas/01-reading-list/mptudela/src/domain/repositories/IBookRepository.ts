import { Book } from '@/domain/models/Book';

export interface IBookRepository {
    getBooks: () => Book[];
}
