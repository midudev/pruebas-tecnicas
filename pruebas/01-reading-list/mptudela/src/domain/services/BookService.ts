import { IBookRepository } from '@/domain/repositories/IBookRepository';
import { Book } from '@/domain/models/Book';
import { SearchParams } from '@/domain/models/SearchParams';

export class BookService {
    private bookRepository: IBookRepository;
    constructor(bookRepository: IBookRepository) {
        this.bookRepository = bookRepository;
    }
    getAllBooks() {
        return this.bookRepository.getBooks();
    }
    filterByTitle(book: Book, title: string) {
        if (!title) return book;
        return book.title.toLowerCase().includes(title.toLowerCase());
    }
    filterByGenre(book: Book, genres: string[]) {
        if (!genres.length) return book;
        return genres.some(genre => book.genre.includes(genre));
    }
    filterByMaxPages(book: Book, maxPages: number) {
        if (maxPages === 0) return book;
        return book.pages <= maxPages;
    }
    filterBooks(books: Book[], searchParams: SearchParams) {
        return books
            .filter(
                book =>
                    this.filterByTitle(book, searchParams.title) &&
                    this.filterByGenre(book, searchParams.genres) &&
                    this.filterByMaxPages(book, searchParams.maxPages)
            )
            .sort((a, b) => a.title.localeCompare(b.title));
    }
}
