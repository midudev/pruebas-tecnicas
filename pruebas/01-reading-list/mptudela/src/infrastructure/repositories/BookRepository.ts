import { Book } from '@/domain/models/Book';
import { IBookRepository } from '@/domain/repositories/IBookRepository';
import data from '@/mocks/books.json';

export class BookRepository implements IBookRepository {
    getBooks() {
        return data.library.map(({ book }) => book as Book);
    }
}
