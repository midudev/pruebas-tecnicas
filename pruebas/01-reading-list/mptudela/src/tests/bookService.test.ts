import { BookService } from '@/domain/services/BookService';
import { BookRepository } from '@/infrastructure/repositories/BookRepository';
import { assert, describe, expect, it } from 'vitest';

const books = [
    {
        title: 'El Señor de los Anillos',
        pages: 1200,
        genre: 'Fantasía',
        cover: 'https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1566425108i/33.jpg',
        synopsis: 'Una aventura épica en un mundo de fantasía llamado la Tierra Media.',
        year: 1954,
        ISBN: '978-0618640157',
        author: {
            name: 'J.R.R. Tolkien',
            otherBooks: ['El Hobbit', 'El Silmarillion']
        }
    },
    {
        title: 'Juego de Tronos',
        pages: 694,
        genre: 'Fantasía',
        cover: 'https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1273763400i/8189620.jpg',
        synopsis: 'En un reino donde las estaciones duran años, una batalla épica por el trono se desarrolla.',
        year: 1996,
        ISBN: '978-0553103540',
        author: {
            name: 'George R. R. Martin',
            otherBooks: ['Choque de Reyes', 'Tormenta de Espadas', 'Festín de Cuervos']
        }
    },
    {
        title: 'Harry Potter y la piedra filosofal',
        pages: 223,
        genre: 'Fantasía',
        cover: 'https://images-na.ssl-images-amazon.com/images/S/compressed.photo.goodreads.com/books/1550337333i/15868.jpg',
        synopsis: 'Un niño descubre que es un mago y comienza una aventura en una escuela de magia.',
        year: 1997,
        ISBN: '978-0747532699',
        author: {
            name: 'J.K. Rowling',
            otherBooks: ['Harry Potter y la cámara secreta', 'Harry Potter y el prisionero de Azkaban', 'Harry Potter y el cáliz de fuego']
        }
    }
];

describe('bookService', () => {
    const bookRepository = new BookRepository();
    const bookService = new BookService(bookRepository);

    it('should filter a list of books and return an array of results', () => {
        const searchParams = { title: 'harry', genres: ['Fantasía'], maxPages: 300 };
        const filteredBooks = bookService.filterBooks(books, searchParams);
        assert(filteredBooks.length === 1);
        expect(filteredBooks[0].title).toBe('Harry Potter y la piedra filosofal');
    });
    it('should return an empty array if no books match the search params', () => {
        const searchParams = { title: 'harry', genres: ['Fantasía'], maxPages: 100 };
        const filteredBooks = bookService.filterBooks(books, searchParams);
        assert(filteredBooks.length === 0);
    });
    it('should return all books if no search params are provided', () => {
        const searchParams = { title: '', genres: [], maxPages: 0 };
        const filteredBooks = bookService.filterBooks(books, searchParams);
        assert(filteredBooks.length === 3);
    });
});
