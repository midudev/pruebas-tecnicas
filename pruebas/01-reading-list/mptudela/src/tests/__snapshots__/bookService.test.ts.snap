// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`suite name > snapshot 1`] = `
{
  "foo": "bar",
}
`;
